### PR TITLE
Improve discriminated union

### PR DESCRIPTION
**Pull request by colinhacks**

### CODE REVIEW

File: deno/lib/__tests__/discriminatedUnions.test.ts [hunk 1]
```diff
@@ -217,3 +229,215 @@ test("valid - literals with .default or .preprocess", () => {
     a: "foo",
   });
 });
+
+test("valid - `DiscriminatedUnion` as a union option", () => {
+  const A = z.object({ type: z.literal("a"), a: z.literal(1) });
+  const B = z.object({ type: z.literal("b"), b: z.literal(2) });
+  const C = z.object({ type: z.literal("c").optional(), c: z.literal(true) });
+  const AorB = z.discriminatedUnion("type", [A, B]);
+  const schema = z.discriminatedUnion("type", [AorB, C]);
```
Comment (jamesarosen) [inline deno/lib/__tests__/discriminatedUnions.test.ts:238]: 👏🏼

**Comment by mharj** [comment]
any progress to merge this?

**Comment by maheshbansod** [comment]
Nested discriminated unions are quite nice.
Since this hasn't been merged, are there any workarounds for this?

**Comment by ghimisradu** [comment]
Will this get merged, or are there any workarounds for nested discriminated unions?

**Comment by thiagozf** [comment]
I just came across this issue. Is there something blocking the PR? I'd be happy to contribute.

**Comment by dani-beltran** [comment]
Why is this being ignored ? This is needed feature.

**Comment by dani-beltran** [comment]
I see this feature is added finally in v4 👍 
Thanks

**Pull request closed without merging**